python-squid/src/main/java/org/sonar/python/metrics/CognitiveComplexityVisitor.java

```diff
@@ -0,0 +1,232 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.metrics;
+
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.AstNodeType;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.sonar.python.PythonVisitor;
+import org.sonar.python.api.PythonGrammar;
+import org.sonar.python.api.PythonKeyword;
+import org.sonar.python.api.PythonPunctuator;
+
+public class CognitiveComplexityVisitor extends PythonVisitor {
+
+  private int complexity = 0;
+  private Deque<NestingLevel> nestingLevelStack = new LinkedList<>();
+
+  @Nullable
+  private final SecondaryLocationConsumer secondaryLocationConsumer;
+
+  public interface SecondaryLocationConsumer {
+    void consume(AstNode node, String message);
+  }
+
+  CognitiveComplexityVisitor(@Nullable SecondaryLocationConsumer secondaryLocationConsumer) {
+    this.secondaryLocationConsumer = secondaryLocationConsumer;
+    nestingLevelStack.push(new NestingLevel());
+  }
+
+  public static int complexity(AstNode node, @Nullable SecondaryLocationConsumer secondaryLocationConsumer) {
+    CognitiveComplexityVisitor visitor = new CognitiveComplexityVisitor(secondaryLocationConsumer);
+    visitor.scanNode(node);
+    return visitor.complexity;
+  }
+
+  public int getComplexity() {
+    return complexity;
+  }
+
+  @Override
+  public Set<AstNodeType> subscribedKinds() {
+    return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+      PythonGrammar.IF_STMT,
+      PythonKeyword.ELIF,
+      PythonKeyword.ELSE,
+
+      PythonGrammar.WHILE_STMT,
+      PythonGrammar.FOR_STMT,
+      PythonGrammar.EXCEPT_CLAUSE,
+
+      PythonGrammar.AND_TEST,
+      PythonGrammar.OR_TEST,
+
+      PythonGrammar.TEST,
+
+      PythonGrammar.FUNCDEF,
+      PythonGrammar.CLASSDEF,
+      PythonGrammar.SUITE)));
+  }
+
+  @Override
+  public void visitNode(AstNode astNode) {
+    if (astNode.is(PythonGrammar.FUNCDEF, PythonGrammar.CLASSDEF)) {
+      nestingLevelStack.push(new NestingLevel(nestingLevelStack.peek(), astNode));
+    } else if (astNode.is(PythonGrammar.SUITE)) {
+      if (isSuiteIncrementsNestingLevel(astNode)) {
```

Comment:
You could merge this inner if condition with its parent one, as all following ifs are also testing node type

Reply:
It was intentional to separate conditions and end the `else if` chain for `astNode.is(PythonGrammar.SUITE)` here. I see it as: easier to read and better performance.